docs/Protected_build_porting_guide.md

```diff
@@ -0,0 +1,363 @@
+- Secondly, initialize and configure the MPU.
+- A sample is shown in the code snippet below :
+
+```
+void CHIPSET_mpuinitialize(void)
+{
+	....
+	for (i = 0; i < (sizeof(regions_info) / sizeof(struct mpu_region_info)); i++) {
+		lldbg("Region = %u base = 0x%x size = %u\n", regions_info[i].rgno, regions_info[i].base, regions_info[i].size);
+		regions_info[i].call(regions_info[i].rgno, regions_info[i].base, regions_info[i].size);
+	}
+	....
+
+	mpu_control(true, false, true);		/*ENABLE the MPU : this enables priviledged access to default memory map */
+}
+```
+
+- This API is to be called by '__start()' during boot up phase.
+
+
+
+
+### Allocation of kernel and user heaps
+- For allocation of memory for kernel and user heaps, create a file 'CHIPSET_allocateheap.c' at path <os/arch/arm/src/CHIPSET>
+- For example, the below given API is written to for kernel-space heap allocation:
+
+```
+	#if defined(CONFIG_BUILD_PROTECTED) && defined(CONFIG_MM_KERNEL_HEAP)
+	void up_allocate_kheap(FAR void **heap_start, size_t *heap_size)
+	{
+		//The kernel heap will start at the end of the idle stack
+		*heap_start = (FAR void *)(g_idle_topstack & ~(0x7));
+
+		//The kernel heap will extend from the end of idle stack to the start of user ram region
+		//The start address of the user ram region is obtained from the variables defined in memory.ld linker script
+		*heap_size = (uint32_t)((uintptr_t)__usram_segment_start__) - (uint32_t)(*heap_start);
+	}
+	#endif
+```
+
+- Similarly, for user-space heap allocation the following API is written. This function provides the size of the unprotected, user-space heap.
+
+
+```
+	void up_allocate_heap(FAR void **heap_start, size_t *heap_size)
+	{
+	#if defined(CONFIG_BUILD_PROTECTED) && defined(CONFIG_MM_KERNEL_HEAP)
+
+		//The user heap will extend from the end of bss section to the end of user ram region
+		// The start address and the size of the user ram region is stored in variables defined in memory.ld script
+		uintptr_t user_end = (uint32_t)__usram_segment_start__ + (uint32_t)__usram_segment_size__;
+		//The user heap will start at the end of the .bss section
+		uintptr_t ubase = (uintptr_t)USERSPACE->us_bssend;
+		size_t usize = (uint32_t)user_end - ubase;
+		....
+	#endif
+	}
+```
+
+- Both these APIs are called at the time of the OS boot-up by the 'os_start.c' file (path: <os/kernel/init>)
+- It is mandatory that API signature shown in the above snippets is not changed.
+
+
+
+### Changes in Make.defs and Makefile
+- The above created files need to be included in the Makefile for compilation during the build process.
+- Add these files to the macro: CHIP_CSRCS in Make.defs which will be used during the build process by Makefile.
+- The following additions are made in the Make.defs at path <os/arch/arm/src/CHIPSET/Make.defs>
+
+```
+	ifeq ($(CONFIG_BUILD_PROTECTED),y)
+	CHIP_CSRCS += CHIPSET_userspace.c CHIPSET_mpuinit.c CHIPSET_allocateheap.c
```

Comment:
I think `CHIPSET_allocateheap.c` is needed not only protected build, but also flat build.

Reply:
Changes made